src/cache_cf.cc

```diff
@@ -3832,6 +3832,16 @@ parsePortCfg(AnyP::PortCfgPointer *head, const char *optionName)
         }
     }
 
+    if (s->secure.encryptTransport) {
+        // protocol requires TLS, that means cert= is a required parameter
```

Comment:
Actually, peeking at and splicing/terminating TLS connections does not require a certificate. We probably require one today for historical reasons and because many admins want Squid to return error responses before terminating connections, but it is a bug or lack of a feature. I suggest replacing this comment with something more accurate like `no support for certificate-free TLS inspection and splicing/termination (yet)`.

Reply:
well, "no support" is not helpful for admin in fixing the issue. The reason is more specifically cert= being missing, because lack of it causes the SEGV this PR is about while copying an uninitialized shared_ptr with the "signingCa = certs.front()" line. Whether we ever use that signingCa or not is irrelevant to any of these layers of code - which is kind of why it snuck past us in the first place.

Consider the upcoming PR 140. It changes the invariant not to be about cert= anymore - but this same copy line will still crash if it is allowed to happen. It was testing for that PR which found this issue in the main code.

Reply:
> "no support" is not helpful for admin (and us) in fixing the issue.

It is key information IMO. Segfaults change and may even disappear. The understanding of the architectural limitations/invariants is where one should start when trying to remove a given check. That kind of information is often difficult to discover.

However, if you think the suggested replacement is not helpful, I am not going to waste more time arguing about it. You can fix the inaccurate comment some other way. Also, as I said, I can withdraw my polishing request if you insist that the current comment should be committed as is.

Reply:
Hmm, I think our understanding of the invariant is different. IMO the invariant here is that TLS encrypt on a socket/port requires a cert. The reasons for that requirement are elsewhere at the point(s) that flag is set and used.

Note the outer if-condition on 'crypting excludes things like splice/terminate (which do not encrypt or decrypt), and the situation being documented is not restricted (or particularly related to) inspection and such. It is a natural limit that applies to all protocols supported in Squid, and/any "modes" of traffic they contain. They either encrypt/decrypt (cert is _required_) or they don't (cert is _optional_).

Reply:
> IMO the invariant here is that TLS encrypt on a socket/port requires a cert.

AFAICT, the `secure.encryptTransport` condition in question is not about doing TLS encryption (which would require a certificate indeed). The condition is more general. It is about dealing with an https_port configuration, including use cases where no encryption/certificate is used/needed.

> Note the outer if-condition on 'crypting excludes things like splice/terminate

Which one? The `s->secure.encryptTransport` condition I am discussing is the outer if-condition, and it excludes only `http_port` directives AFAICT (i.e., it limits us to `https_port` configurations). In other words, AFAICT, encryptTransport does not mean "we are going to encrypt something". It means "we are going to deal with TLS connections".

Reply:
Sorry misspelling there - I meant "other" as in the "if (s->transport.protocol == AnyP::PROTO_HTTPS)" at line 3799, above the new one. Anyhow my point being that the flag is set based on port type being HTTPS or having TLS context related options configured. The "ssl-bump" option is parsed outside the ServerOptions class and "http_port ssl-bump" does not require a certificate. I have just confirmed that with a-k parse test, though I'm not sure how far into the SSL-Bump processing of actual CONNECT it can get before encountering an error.

If we cant agree on a wording I'm inclined to remove the comment entirely and let the code speak for itself. Would that be sufficient to get a pass vote?

Reply:
_If_ the new FATAL line is not specific to https_port, then perhaps we should replace `..._port requires` with something like `your ..._port configuration requires` because http_port itself certainly does not require a certificate.

> If we cant agree on a wording I'm inclined to remove the comment entirely and let the code speak for itself. Would that be sufficient to get a pass vote?

Yes, provided my understanding of this PR does not change. I was under impression that you are only modifying how https_port configuration is handled (to prevent a segfault). If I had an `http_port ssl-bump` configuration that was "working" without a certificate (because it never triggered/reported errors), will my configuration stop working after this PR?

Reply:
It started there yes. Moving the check away from the context init widened the range of affected configs considerably.

If your port was trying to do things like set tls-ciphers=, or ssl-flags= I expect it would start getting the message or the crash - because those options are about the TLS context which (currently) needs a cert to create. Otherwise no change.

Reply:
> Otherwise no change.

Sounds good.